### PR TITLE
Implemented font color support in front-end and back-end

### LINKED ISSUE #3554: Wizard Step 3 - Session Colors: Show preview and automatically choose black or white font color
Currently we only have standard font color white or black in the list and calendar views of the schedule.

The font color should be chosen automatically depending on the contrast. Please implement the following:
* behind each line implement a preview area with the background and automatically chosen font color a) black or b) white
* implement the font color in the schedule view
* add the font color choice to the API and JSON

screenshot from 2017-04-28 11-19-12
Show a small preview on the right here.

There is a slack discussion on this feature here http://stackoverflow.com/questions/1855884/determine-font-color-based-on-background-color/1855903

### PR DESCRIPTION
**Pull request by mejariamol**
fixes #3554 

screenshot from 2017-05-06 01-42-06
screenshot from 2017-05-06 01-44-58

### CODE REVIEW

File: app/templates/gentelella/users/events/wizard/components/tracks-microlocations-sessions.html [hunk 1]
```diff
@@ -1,32 +1,35 @@
...
-        <div class=" col-xs-5 form-group">
+        <div class=" col-xs-4 form-group">
             <input type="text" class="form-control" v-model="track.name" placeholder="Name" required="required">
         </div>
-        <div class="col-xs-7 input-group form-group">
+        <div class="col-xs-4 input-group form-group" style="float:left;">
             <div class="input-group colorpicker-component" id="color">
                 <input type="text" class="form-control color-picker-input" placeholder="Color" required="required"
-                       v-model="track.color" title="track-color"/>
+                       v-model="track.color" title="track-color" v-on:input="changefont"/>
```
Comment (niranjan94) [inline app/templates/gentelella/users/events/wizard/components/tracks-microlocations-sessions.html]: Use a watcher on `track.color` to trigger the font color change

File: app/static/js/events/wizard/sessions-speakers/vue-components.js [hunk 2]
```diff
@@ -7,6 +7,9 @@ Vue.component('session-track', {
         },
         add: function () {
             this.$emit('add');
+        },
+        changefont: function () {
```
Comment (niranjan94) [inline app/static/js/events/wizard/sessions-speakers/vue-components.js]: Why are you emitting this ?

File: app/api/sessions.py [hunk 3]
```diff
@@ -31,6 +31,7 @@
     'id': fields.Integer(required=True),
     'name': fields.String(),
     'color': fields.Color(),
+    'font_color': fields.Color(),
```
Comment (niranjan94) [inline app/api/sessions.py]: Don't allow user to set font color. Font color needs to be a readonly attribute on the API

File: app/models/track.py [hunk 4]
```diff
@@ -9,19 +9,21 @@ class Track(db.Model):
     description = db.Column(db.Text)
     track_image_url = db.Column(db.Text)
     color = db.Column(db.String, nullable=False)
+    font_color = db.Column(db.String)
```
Comment (niranjan94) [inline app/models/track.py]: Don't store fontcolor to data base. It's a generated property. Generate it based on track color and provide it in the API

File: app/api/tracks.py [hunk 5]
```diff
@@ -24,6 +24,7 @@
     'name': fields.String(required=True),
     'description': fields.String(),
     'color': fields.Color(required=True),
+    'font_color': fields.Color(required=True),
```
Comment (niranjan94) [inline app/api/tracks.py]: Same comment as above.

File: app/static/js/events/wizard/sessions-speakers/vue-components.js (hunk 2 above)
Comment (mejariamol) [inline app/static/js/events/wizard/sessions-speakers/vue-components.js]: child component has to emit the event for parent to receive right? Ok I will go for *watch implementation*.

File: app/models/track.py (hunk 4 above)
Comment (mejariamol) [inline app/models/track.py]: @niranjan94 how can we provide font color in the api without saving in database ?

Comment (niranjan94) [inline app/models/track.py]: Define a property in the track model (in `app/models/track.py`). Similar to the already existing `serialize` property. Then you can just use it like a db field in the API Model definition.

Comment (mejariamol) [inline app/models/track.py]: @niranjan94 ok.. so that means there will be a function in `app/models/track.py`  generating fontcolor from color right?

**Comment by niranjan94** [comment]
Yes. Correct.

**Comment by mejariamol** [comment]
@niranjan94 have a look at last changes.. 

1. font_color is defined as property in `app/models/track.py`. 
2. set watcher to **color** to change font_color
3. but couldn't understand how to get font_color field in the get request only in api .

**Comment by niranjan94** [comment]
@mejariamol looks good. You should already get font color in the API now. 
All you have to do next is to make sure the API doesn't try to save font_color when someone makes a post/put/patch request to tracks endpoint

**Comment by shubham-padia** [comment]
@mejariamol Try adding `del TRACK_POST['font_color']` on https://github.com/mejariamol/open-event-orga-server/blob/a9ba40a0275c5b2060eb481bef13237a8b22d0a2/app/api/tracks.py#L40, AFAIK that should make the field read-only.

**Comment by mejariamol** [comment]
@niranjan94 @shubham-padia Travis CI is failing due to testing of color field against `red` in the function where I am parsing r g b from hex code. `int(h[0:2], 16)`.

**Comment by niranjan94** [comment]
@mejariamol you need to add font color here https://github.com/fossasia/open-event-orga-server/blob/development/tests/unittests/api/utils_post_data.py#L118

**Comment by mejariamol** [comment]
@niranjan94 yes. but then also it will fail while running this.. right?
```
@property
    def font_color(self):
        h = self.color.lstrip('#')
        a = 1 - (0.299 * int(h[0:2], 16) + 0.587 * int(h[2:4], 16) + 0.114 * int(h[4:6], 16))/255
        ...
```

**Comment by mejariamol** [comment]
@niranjan94 travis build passed for last commit. can be merged now.

**Pull request merged**